File: security/prescription_guard.go
package security

import (
	"context"
	"net/http"

	"doctor-service/entities"
	"doctor-service/repositories"

	"github.com/google/uuid"
)

type PrescriptionAuthGuard struct {
	doctorRepo repositories.DoctorRepository
}

func NewPrescriptionAuthGuard(doctorRepo repositories.DoctorRepository) *PrescriptionAuthGuard {
	return &PrescriptionAuthGuard{doctorRepo: doctorRepo}
}

func (g *PrescriptionAuthGuard) RequireVerifiedDoctorForPrescription(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		docIDStr := r.Header.Get("X-Doctor-Public-ID")
		if docIDStr == "" {
			http.Error(w, `{"error":"Unauthorized: Missing X-Doctor-Public-ID header"}`, http.StatusUnauthorized)
			return
		}

		docUUID, err := uuid.Parse(docIDStr)
		if err != nil {
			http.Error(w, `{"error":"Unauthorized: Invalid X-Doctor-Public-ID header format"}`, http.StatusUnauthorized)
			return
		}

		doc, err := g.doctorRepo.FindByPublicID(r.Context(), docUUID)
		if err != nil || doc == nil {
			http.Error(w, `{"error":"Unauthorized: Doctor account not found"}`, http.StatusUnauthorized)
			return
		}

		if doc.Status != entities.DoctorStatusVerified {
			http.Error(w, `{"error":"Forbidden: Doctor verification incomplete. Only VERIFIED doctors can generate digital prescriptions or conduct teleconsultations."}`, http.StatusForbidden)
			return
		}

		if !doc.PrescriptionEnabled {
			http.Error(w, `{"error":"Forbidden: Digital prescription enablement is disabled on this profile. Please contact medical compliance."}`, http.StatusForbidden)
			return
		}

		// Inject verified doctor into context
		ctx := context.WithValue(r.Context(), "verified_doctor", doc)
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

File: backend_api/security/prescription_guard.py
import json
from uuid import UUID

class PrescriptionAuthGuard:
    def __init__(self, doctor_repo):
        self.doctor_repo = doctor_repo

    def require_verified_doctor_for_prescription(self, request, next_handler):
        doc_id_str = request.headers.get("X-Doctor-Public-ID")
        if not doc_id_str:
            return {"error": "Unauthorized: Missing X-Doctor-Public-ID header"}, 401
        
        try:
            doc_uuid = UUID(doc_id_str)
        except ValueError:
            return {"error": "Unauthorized: Invalid X-Doctor-Public-ID header format"}, 401
            
        doc = self.doctor_repo.find_by_public_id(request.context, doc_uuid)
        if not doc:
            return {"error": "Unauthorized: Doctor account not found"}, 401
            
        if doc.status != "VERIFIED":
            return {"error": "Forbidden: Doctor verification incomplete. Only VERIFIED doctors can generate digital prescriptions or conduct teleconsultations."}, 403
            
        if not doc.prescription_enabled:
            return {"error": "Forbidden: Digital prescription enablement is disabled on this profile. Please contact medical compliance."}, 403
            
        request.context["verified_doctor"] = doc
        return next_handler(request)
